submissions/factorio_draftsman/main.md

```diff
@@ -0,0 +1,200 @@
+            index = 0
+
+    # Add the last combinator if partially full
+    combinator.id = "{}_{}".format(x, y)
+    blueprint.entities.append(combinator)
+
+    # Add connections to each neighbour
+    for cx in range(x):
+        for cy in range(5):
+            here = "{}_{}".format(cx, cy)
+            right = "{}_{}".format(cx + 1, cy)
+            below = "{}_{}".format(cx, cy + 1)
+            try:
+                blueprint.add_circuit_connection("red", here, right)
+            except KeyError:
+                pass
+            try:
+                blueprint.add_circuit_connection("red", here, below)
+            except KeyError:
+                pass
+
+    print("Number of items:", count)
+    print(blueprint.to_string())
+
+
+if __name__ == "__main__":
+    main()
+```
+
+This script is concise and easy to follow, but the really amazing thing about Draftsman is that this script is completely dynamic and can adjust to whatever version of Factorio that you're playing. New items, removed items, or changed stack sizes, regardless of the author of the change, are resolved automatically with the exact same code. For illustration, pictured top-left is what's generated by running the script for a vanilla game, top-right a game with the medium-sized overhaul modpack [Space Exploration](https://mods.factorio.com/mod/space-exploration), and at the bottom a [Bobs](https://mods.factorio.com/user/Bobingabout) + [Pyanodon](https://mods.factorio.com/user/pyanodon) mega-modpack:
+
+![Results](./img/item_stack_size_output.png)
+
+## Image to Blueprint Converter
+
+[This was something I made on a whim.](https://github.com/redruin1/factorio-draftsman/blob/main/examples/image_converter.py) It uses Pillow to load an image and converts it to a blueprint intended to be visible from the map view, all in less than 150 lines of code:
+
+![Portrait of Mike](./img/mike_stoklasa.png "A drunk bum with comically large flask, colorized.")
+
+Many improvements could be made to this:
+
+* Dithering is only implemented on 1x1 tiles; I was having trouble adjusting the algorithm to multi-tile entities.
+* Multi-tile entities also fail to adjust the error metric property; palettization doesn't work as well when some pixels are of different size.
+* Some multi-tile entities have different rotations; a better implementation would check which orientation produces the least error before placing, instead of only placing them with the default orientation.
+* The colors themselves are hardcoded; it would be nice to dynamically load this from the game, especially with modded map colors...
+
+Despite these shortcomings, the output actually ends up pretty good for it's quick and dirty implementation. This shows that Draftsman is versatile enough for some rather eclectic purposes.
+
+![Portrait of Jay](./img/jay_bauman.png "Jay Bauman, notorious hack fraud, criminal at large. Armed and dangerous.")
+
+## Factorio Movie Player Resurrection
```

Comment:
Ah man this is epic!

Reply:
https://www.youtube.com/watch?v=jrHA6JlrOHE